earth_enterprise/src/fusion/autoingest/AssetRegistry.h

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2020 The Open GEE Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef ASSET_REGISTRY_H
+#define ASSET_REGISTRY_H
+
+#include <functional>
+#include <memory>
+#include <unordered_map>
+
+template <class AssetType>
+class AssetRegistry{
+public:
+  AssetRegistry() = delete;
+
+  class AssetPluginInterface {
+    /*
+    * Collection of plugin-specific versions of functions.
+    */
+    public:
+      using new_from_dom_func = std::function<std::shared_ptr<AssetType>(void*)>;
+      using new_invalid_func = std::function<std::shared_ptr<AssetType>(const std::string &ref)>;
+      new_from_dom_func pNewFromDom;
+      new_invalid_func pNewInvalid;
+
+      explicit AssetPluginInterface(new_from_dom_func pNFD, new_invalid_func pNI):
+        pNewFromDom(pNFD), pNewInvalid(pNI) {}
+  };
+
+
+  class PluginRegistrar
+  {
+    /*
+    * PluginRegistrar objects are instantiated within a plugin's cpp file to
+    * register the plugin.
+    */
+    public:
+      explicit PluginRegistrar(
+        std::string const & name, 
+        std::unique_ptr<AssetPluginInterface> assetPlugin)
+      {
+          AssetRegistry<AssetType>::RegisterPlugin(name, std::move(assetPlugin));
+      }
+  };
+
+  static void RegisterPlugin(
+    std::string const & assetTypeName, 
+    std::unique_ptr<AssetPluginInterface> assetPlugin)
+  {
+    PluginRegistry()[assetTypeName] = std::move(assetPlugin);
+  }
+
+  static AssetPluginInterface * GetPlugin(const std::string & assetTypeName) {
+    auto it = PluginRegistry().find(assetTypeName);
+    if (it != PluginRegistry().end())
```

Comment:
If this if statement fails it means something is wrong with the code, correct? We should never have a plugin that hasn't registered its interface. Maybe we can change this statement into an assert and always return `it->second.get()`. That way, code that calls this function doesn't have to check for `nullptr`.

Reply:
I take it back. You need the if/else statement in case there's a malformed XML file. It may be worth considering if there's a way to avoid returning nullptr.

Reply:
Malformed XML could cause a failure in `AssetFactory::CreateNewFromDOM`, but I don't think it comes into play here. 
I can't think of any condition where this could/should return `nullptr`. Unless an asset type was removed from one version of Fusion to another, but I can't imagine that happening.

Reply:
Are you guaranteed to have a valid `assetTypeName` at this point? I thought it was read straight from the XML. So maybe I'm not so much concerned about malformed XML as well-formed XML with garbage data.

Reply:
"Guaranteed" is probably a strong word. It's coming from an XML file that Fusion created, so it's probably _relatively_ safe except in the case of a corrupted file, but I'm all for leaving the check in there as good practice.